### PR TITLE
Resource status

### PR DESCRIPTION
**Pull request by thefron**
Display book action button based on resource status.
We need to address N+1 request later on.

Screen Shot 2019-07-01 at 10 34 51 AM

### CODE REVIEW

File: frontend/src/components/BookListCard/BookListCard.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    book.setName("Test Book");
+    book.setAuthorsList(["Me", "Myself", "I"]);
+    book.setCoverImage("https://placekitten.com/g/200/300");
+    return book;
+})();
+
+storiesOf("Components|BookListCard", module)
+  .add("default", () => (
+    <MuiThemeProvider theme={StandardTheme}>
+      <BookListCard book={TestBook}/>
```
Comment (Goblinlordx) [inline frontend/src/components/BookListCard/BookListCard.stories.tsx]: This works?

Comment (thefron) [reply frontend/src/components/BookListCard/BookListCard.stories.tsx]: @Goblinlordx Actually no. I think `proto` directory is not being resolved properly..

Comment (Goblinlordx) [reply frontend/src/components/BookListCard/BookListCard.stories.tsx]: It looks like there are props missing?  Maybe you can inject a mock for this like is done for TestBook?

Comment (thefron) [reply frontend/src/components/BookListCard/BookListCard.stories.tsx]: Yes I'm aware of that, but it cannot resolve rentalsvc pb properly before that.

Comment (thefron) [reply frontend/src/components/BookListCard/BookListCard.stories.tsx]: @Goblinlordx After fixing storybook webpack alias path, it's being rendered in storybook!

File: frontend/src/components/BookListCard/BookListCard.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+const RatingLabel: any = styled(Typography)`
+  text-align: middle;
+  padding: 0 8px 0 4px;
+`;
+
+interface BookListCardInterface {
+  book: Book
+  currentUser?: User
+  statusFetcher?: (bookId: number) => [boolean, ResourceStatus]
```
Comment (Goblinlordx) [inline frontend/src/components/BookListCard/BookListCard.tsx]: What if the fetch fails?  More importantly, what is it returning before it is fetched?

File: frontend/src/hooks/RentalStatus/RentalStatus.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        .then(status => {
+          if (cancelled()) return;
+          setStatus(status);
+        })
+      );
+      return cancel;
+  };
+
+  useEffect(getResourceStatus(bookId), []);
+  return [loading, status || new ResourceStatus];
```
Comment (Goblinlordx) [inline frontend/src/hooks/RentalStatus/RentalStatus.tsx]: Why is this returning something that it doesn't have?

File: frontend/src/hooks/RentalStatus/RentalStatus.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+const rentalService = new RentalServiceClient(process.env.PUBLIC_URL, null, null);
+
+const getResourceStatus = (bookId: number): [boolean, ResourceStatus] => {
+  const [loading, load] = Loader();
+  const [status, setStatus] = useState<ResourceStatus>();
+  const getResourceStatus = (bookId: number) => () => {
+      const req = new GetResourceStatusRequest();
+      req.setEntityId(bookId)
+      const statusPromise: Promise<ResourceStatus> = new Promise((resolve, reject) => {
+        rentalService.getResourceStatus(req, {}, (err: any, status: ResourceStatus) => err ? reject(err) : resolve(status))
```
Comment (Goblinlordx) [inline frontend/src/hooks/RentalStatus/RentalStatus.tsx]: Try not to shadow variables.

File: frontend/src/hooks/RentalStatus/RentalStatus.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import { useState, useEffect } from "react";
+import { RentalServiceClient } from 'proto/rentalsvc_grpc_web_pb';
+import { GetResourceStatusRequest, ResourceStatus } from 'proto/rentalsvc_pb';
+import Loader from "../Loader/Loader";
+
+const rentalService = new RentalServiceClient(process.env.PUBLIC_URL, null, null);
+
+const getResourceStatus = (bookId: number): [boolean, ResourceStatus] => {
+  const [loading, load] = Loader();
+  const [status, setStatus] = useState<ResourceStatus>();
```
Comment (Goblinlordx) [inline frontend/src/hooks/RentalStatus/RentalStatus.tsx]: This type does not seem correct as it is initialized with `undefined`.

File: frontend/src/hooks/RentalStatus/RentalStatus.tsx [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  const getResourceStatus = (bookId: number) => () => {
+      const req = new GetResourceStatusRequest();
+      req.setEntityId(bookId)
+      const statusPromise: Promise<ResourceStatus> = new Promise((resolve, reject) => {
+        rentalService.getResourceStatus(req, {}, (err: any, status: ResourceStatus) => err ? reject(err) : resolve(status))
+      });
+
+      const [cancelled, cancel] = load(statusPromise
+        .then(status => {
```
Comment (Goblinlordx) [inline frontend/src/hooks/RentalStatus/RentalStatus.tsx]: Same here with shadowing.  Might have missed something like this before =/.

**Pull request merged**